### PR TITLE
fix issue 16072 - container.binaryheap should be extendable for arrays

### PR DESCRIPTION
**Pull request by wilzbach**
While trying to create an top-level unittest for this module (#4331), I realized that with an array as storage, it's not possible to insert new items, which seems like a bug. 

```
import std.container.binaryheap;
auto q = heapify!"a > b"([1, 4, 5]);
q.insert(2); // ERROR: Cannot grow a heap
```

Here's the fix ;-)

### CODE REVIEW

**Comment by ZombineDev** [comment]
Can you add a unittest? Otherwise, LGTM.

**Comment by aG0aep6G** [comment]
Maybe `insertBack` should work on arrays instead?

**Comment by wilzbach** [comment]
> Can you add a unittest? Otherwise, LGTM.

Done. Also added a check + fix for the case of a zero-length array.

> Maybe insertBack should work on arrays instead?

Afaict is `insertBack` deliberately picked for the container API to avoid interferring with range API.

See also: http://dlang.org/phobos-prerelease/std_container.html

**Comment by aG0aep6G** [comment]
> Afaict is insertBack deliberately picked for the container API to avoid interferring with range API.

I don't see how they would interfere. Arrays already support a bunch of the container primitives. Adding the rest seems more natural to me than special casing function that work on containers.

**Comment by wilzbach** [comment]
> I don't see how they would interfere. Arrays already support a bunch of the container primitives. Adding the rest seems more natural to me than special casing functions that work on containers.

I actually asked the same question before and IIRC it was about guareenting runtime based on the name (?). Ping @schveiguy

**Comment by schveiguy** [comment]
@wilzbach I have no experience with binary heap, so I don't really have any opinion on it.

**Comment by wilzbach** [comment]
> @wilzbach I have no experience with binary heap, so I don't really have any opinion on it.

I thought the container API is similiar between all storages. Sorry.

> I actually asked the same question before

See #4038 - ping @JackStouffer

**Comment by JackStouffer** [comment]
What if store is a slice of an array?

**Comment by schveiguy** [comment]
> What if store is a slice of an array?

You mean a dynamic array? Should work.

**Comment by JackStouffer** [comment]
I guess I misunderstood how _store worked.

File: std/container/binaryheap.d [hunk 1]
```diff
@@ -265,9 +265,20 @@ and $(D length == capacity), throws an exception.
         }
         else
         {
-            // can't grow
```
Comment (JackStouffer) [inline std/container/binaryheap.d]: This change would be cleaner if it was written in the following way

```
        else static if (isDynamicArray!Store)
        {
            if (_store.length == 0)
                _store.length = 10;
            if (length == _store.length)
                _store.length = length * 2;
            _store[_length] = value;
        }
        else
        {
            // can't grow
            enforce(length < _store.length,
                    "Cannot grow a heap created over a range");
        }
```

File: std/container/binaryheap.d [hunk 2]
```diff
@@ -265,9 +265,20 @@ and $(D length == capacity), throws an exception.
...
-            enforce(length < _store.length,
-                    "Cannot grow a heap created over a range");
+            import std.traits: isDynamicArray;
+            static if (isDynamicArray!Store)
+            {
+                if (_store.length == 0)
+                    _store.length = 10;
```
Comment (andralex) [inline std/container/binaryheap.d]: use a power of 2, e.g. 8

File: std/container/binaryheap.d [hunk 3]
```diff
@@ -265,9 +265,20 @@ and $(D length == capacity), throws an exception.
...
+                if (length == _store.length)
+                    _store.length = length * 2;
```
Comment (andralex) [inline std/container/binaryheap.d]: use `length * 3 / 2`

**Comment by andralex** [comment]
OK, interesting... wonder what the consequences of this precedent will be :)

File: std/container/binaryheap.d (hunk 1 above)
Comment (wilzbach) [inline std/container/binaryheap.d]: I didn't know the reason for setting the element after an exception was thrown, so I didn't change the code, but thanks for the heads-up -> changed :)

**Comment by wilzbach** [comment]
> OK, interesting... wonder what the consequences of this precedent will be :)

What consequences do you expect? I think the other containers support insertions ;-)

(rebased)

**Comment by wilzbach** [comment]
>> OK, interesting... wonder what the consequences of this precedent will be :)
> @andralex So is this PR a go or no-go?

I would interpret the "OK, interesting" as an approval, but maybe other diviner are better in interpreting (?)

**Comment by JackStouffer** [comment]
@quickfur I see Andrei's comment as consent

I think this PR is ready to go

**Comment by andralex** [comment]
Auto-merge toggled on

**Pull request merged**